File: app/services/chat_model_settings.py
"""Chat default model + aliases: DB JSON, admin PATCH validation, runtime resolution.

``app_settings.chat_model_aliases`` is a JSON list of ``{alias, provider, model_id}``.
If ``default_chat_model`` equals an alias slug, runtime uses that row's provider and
model_id; else ``model_provider`` + ``default_chat_model`` as the API model id.

Invalid JSON rows are skipped on read; PATCH rejects unknown providers. Writes are
sorted by alias for stable storage. Admin UI mirrors this in
``frontend/src/lib/chatModelSettings.ts`` (PATCH body + dirty fingerprint + provider
for GET payloads). Valid providers: ``MODEL_PROVIDER_OPTIONS`` in ``openai_catalog``.
"""

from __future__ import annotations

import json
from collections.abc import Iterable
from typing import Any

from app.services.openai_catalog import MODEL_PROVIDER_OPTIONS

__all__ = [
    "aliases_from_json_value",
    "normalize_aliases_for_storage",
    "resolve_default_with_aliases",
]


def _coerce_one_row(item: Any) -> dict[str, str] | None:
    if not isinstance(item, dict):
        return None
    alias = str(item.get("alias", "")).strip()
    prov = str(item.get("provider", "")).strip().lower()
    mid = str(item.get("model_id", "")).strip()
    if not alias or not prov or not mid:
        return None
    if prov not in MODEL_PROVIDER_OPTIONS:
        return None
    return {"alias": alias, "provider": prov, "model_id": mid}


def aliases_from_json_value(raw: str | None) -> list[dict[str, str]]:
    if not raw or not str(raw).strip():
        return []
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return []
    if not isinstance(data, list):
        return []
    seen: set[str] = set()
    out: list[dict[str, str]] = []
    for item in data:
        row = _coerce_one_row(item)
        if not row:
            continue
        lk = row["alias"].lower()
        if lk in seen:
            continue
        seen.add(lk)
        out.append(row)
    out.sort(key=lambda r: r["alias"].lower())
    return out


def _triple_from_patch_item(item: Any) -> tuple[str, str, str]:
    if isinstance(item, dict):
        return (
            str(item.get("alias", "")).strip(),
            str(item.get("provider", "")).strip().lower(),
            str(item.get("model_id", "")).strip(),
        )
    alias = str(getattr(item, "alias", "") or "").strip()
    prov = str(getattr(item, "provider", "") or "").strip().lower()
    mid = str(getattr(item, "model_id", "") or "").strip()
    return alias, prov, mid


def normalize_aliases_for_storage(items: Iterable[Any]) -> list[dict[str, str]]:
    seen: set[str] = set()
    out: list[dict[str, str]] = []
    for item in items:
        alias, prov, mid = _triple_from_patch_item(item)
        if not alias or not mid:
            raise ValueError("Each alias needs a non-empty alias and model_id")
        if prov not in MODEL_PROVIDER_OPTIONS:
            raise ValueError(f"Invalid provider for alias {alias!r}: {prov!r}")
        lk = alias.lower()
        if lk in seen:
            continue
        seen.add(lk)
        out.append({"alias": alias, "provider": prov, "model_id": mid})
    out.sort(key=lambda r: r["alias"].lower())
    return out


def resolve_default_with_aliases(
    aliases: list[dict[str, str]],
    *,
    provider: str,
    stored_default_chat_model: str,
) -> tuple[str, str]:
    cm = stored_default_chat_model.strip()
    for a in aliases:
        if a.get("alias") == cm:
            prov = str(a.get("provider", "")).strip().lower()
            mid = str(a.get("model_id", "")).strip()
            if prov in MODEL_PROVIDER_OPTIONS and mid:
                return prov, mid
    p = (provider or "openai").strip()
    if p.lower() == "grok":
        p = "groq"
    return p, cm

File: frontend/src/lib/chatModelSettings.ts
/**
 * Admin chat model UI: PATCH body shape, dirty fingerprint, GET → provider control.
 * Contract: `GET/PATCH /api/settings/config` — fields `model_provider`, `default_chat_model`,
 * `chat_model_aliases`. Server: `app/services/chat_model_settings.py`.
 *
 * `CHAT_PROVIDER_IDS` must match `MODEL_PROVIDER_OPTIONS` in `app/services/openai_catalog.py`.
 */
export const CHAT_PROVIDER_IDS = ["openai", "groq", "openrouter", "huggingface", "nvidia", "tenstorrent"] as const;
export type ChatProviderId = (typeof CHAT_PROVIDER_IDS)[number];

export type ChatModelAliasRow = { alias: string; provider: string; model_id: string };

export type ChatModelConfigSource = {
  model_provider: string;
  default_chat_model: string;
  chat_model_aliases?: ChatModelAliasRow[];
};

export function isChatProviderId(v: string): v is ChatProviderId {
  return (CHAT_PROVIDER_IDS as readonly string[]).includes(v);
}

export function sanitizeChatAliasesForApi(rows: ChatModelAliasRow[]): ChatModelAliasRow[] {
  return rows
    .map((a) => ({
      alias: (a.alias ?? "").trim(),
      provider: (a.provider ?? "").trim().toLowerCase(),
      model_id: (a.model_id ?? "").trim(),
    }))
    .filter((a) => a.alias && a.model_id && isChatProviderId(a.provider))
    .sort((a, b) => a.alias.localeCompare(b.alias));
}

export function chatModelSettingsFingerprint(
  provider: string,
  defaultChatModel: string,
  rows: ChatModelAliasRow[],
): string {
  const p = (provider ?? "").trim().toLowerCase();
  const prov: ChatProviderId = isChatProviderId(p) ? p : "openai";
  const norm = sanitizeChatAliasesForApi(rows);
  return JSON.stringify({ p: prov, m: (defaultChatModel ?? "").trim(), a: norm });
}

export function adminUiChatProvider(c: ChatModelConfigSource): ChatProviderId {
  const rawDm = (c.default_chat_model ?? "").trim();
  const rows = c.chat_model_aliases ?? [];
  const aliasHit = rows.find((a) => a.alias === rawDm);
  const rawProv = (c.model_provider ?? "").trim().toLowerCase();
  if (aliasHit && isChatProviderId((aliasHit.provider ?? "").trim().toLowerCase())) {
    return (aliasHit.provider ?? "").trim().toLowerCase() as ChatProviderId;
  }
  if (rawProv === "grok" || rawProv === "xai") return "groq";
  if (isChatProviderId(rawProv)) return rawProv;
  return "openai";
}

export function buildChatModelSettingsPatchBody(args: {
  provider: ChatProviderId;
  defaultModelId: string;
  aliases: ChatModelAliasRow[];
}): Record<string, unknown> {
  const dm = (args.defaultModelId ?? "").trim();
  return {
    model_provider: args.provider,
    default_chat_model: dm,
    chat_model_aliases: sanitizeChatAliasesForApi(args.aliases),
  };
}
